torch/testing/_internal/opinfo/core.py

```diff
@@ -126,20 +126,44 @@ class SampleInput(object):
     def __init__(
         self,
         input,
-        *,
-        args=tuple(),
+        *var_args,
+        args=None,
         kwargs=None,
         output_process_fn_grad=lambda x: x,
         broadcasts_input=False,
         name="",
+        **var_kwargs,
     ):
         # input is the first input to the op and is typically either a Tensor or TensorList (Sequence[Tensor]).
         # This follows the typical pattern where for Tensor inputs op(t, ...) = t.op(...).
         self.input = input
-        self.args = args
+
+        # Allow calling either as SampleInput(input, args=args, kwargs=kwargs), or as
+        # SampleInput(input, *args, **kwargs) but not to mix the two forms
+        if args is None:
+            self.args = var_args
+        else:
+            assert (
+                len(var_args) == 0
+            ), "SampleInput cannot be constructed with args and var_args"
+            assert (
+                len(var_kwargs) == 0
+            ), "SampleInput cannot be constructed with args and var_kwargs"
+            self.args = args
         assert isinstance(self.args, tuple)
-        self.kwargs = kwargs if kwargs is not None else {}
+
+        if kwargs is None:
+            self.kwargs = var_kwargs
+        else:
+            assert (
```

Comment:
Style suggestion:

```
if args is not None or kwargs is not None:
  assert (len(var_args) == 0 and len(var_kwargs) == 0), "A SampleInput can be constructed "naturally" with *args and **kwargs or by explicitly setting the "args" and "kwargs" paremets, but the two methods of construction cannot be mixed!"

self.args = args if args is not None else var_args
self.kwargs = kwargs if kwargs is not None else var_kwargs
```

So this will still allow developers to set the SampleInput-specific `output_process_fn_grad`, `broadcasts_input`, and `name` parameters inline with function parameters? Per your comment, why not extend this check to disallow those being set if `**kwargs` has values? A message could direct the developer to set them using the fluent helpers you've suggested:

```
# throws an error
SampleInput(a, b, weights=c, output_process_fn=foo)

# works
SampleInput(a, b, weights=c).output_process_fn(foo)
```

cc @pmeier @lezcano if they'd like to help us bikeshed on these UX details

Reply:
+1 on Mikes proposal. I prefer the fluent style, because this makes the `SampleInput` more clear cut.

We could even warn that the old style of passing everything to the constructor is deprecated, but I'm guessing most people will not look at the warnings anyway? If we change most `OpInfo`'s to this style sooner or later most people will start to use it since all references do.